File: Calibration/fit.py
"""
file: fit.py
brief: Script to fit amplitude and/or charge distributions of scintillators
usage: python3 fit.py cfg.yml
note:
author: Alexandre BIGOT, alexandre.bigot@iphc.cnrs.fr
"""

import sys

import ROOT as r

try:
    from yaml import load, FullLoader
except ModuleNotFoundError:
    print("Module 'pyyaml' is not installed. Please install it to run this script.")

try:
    from argparse import ArgumentParser
except ModuleNotFoundError:
    print("Module 'argparse' is not installed. Please install it to run this script.")

try:
    import pandas as pd
except ModuleNotFoundError:
    print("Module 'pandas' is not installed. Please install it to run this script.")

try:
    import uproot
except ModuleNotFoundError:
    print("Module 'uproot' is not installed. Please install it to run this script.")

try:
    sys.path.append("../Utils/")
    from logger import Logger
except ModuleNotFoundError:
    print(
        "Module 'logger' is not in the '../Utils/' directory. Add it to run this script."
    )

try:
    sys.path.append("../Utils/")
    from utils import fill_th1, get_h_config
except ModuleNotFoundError:
    print(
        "Module 'utils' is not in the '../Utils/' directory. Add it to run this script."
    )


# pylint:disable=too-many-locals
def main(name_config_file: str) -> None:
    """
    Main function

    Parameters
    ------------------------------------------------
    - name_config_file: str
        Name of the YAML config file
    """

    # import configuration
    config: dict = {}
    with open(name_config_file, "r", encoding="utf-8") as yml_config_file:
        config = load(yml_config_file, FullLoader)

    # handle input
    name_infile = config["input"]["file"]
    name_tree: str = config["input"]["tree"]["name"]
    name_branches: list[str] = config["input"]["tree"]["branches"]

    h_configs = []
    histogram_cfg = config["histogram_config"]

    # safeties
    if isinstance(name_branches, list) and isinstance(histogram_cfg["name"], list):
        if len(name_branches) != len(histogram_cfg["name"]):
            Logger(
                "'input/branches' and 'histogram_config/name' must be of same size!",
                "FATAL",
            )

    for i, _ in enumerate(name_branches):
        h_configs.append(get_h_config(histogram_cfg, i))

    # convert input .root file into pandas dataframe
    df: pd.DataFrame = uproot.open(name_infile)[name_tree].arrays(
        name_branches, library="pd"
    )

    hists = []
    for name, h_config in zip(name_branches, h_configs):
        hists.append(fill_th1(df[name], h_config))

    # configure output with fit results
    name_outfile: str = config["output"]["name"]
    outfile: r.TFile = r.TFile(name_outfile, "recreate")
    for hist in hists:
        hist.Write()

    # config of fit result storage
    labels_hfit_res: list[str] = [
        "FitStatus",
        "Chi2",
        "NDF",
        "Constant",
        "Mean",
        "Sigma",
        "Xmin",
        "Xmax",
    ]

    hfit_results = []

    # add a fitting procedure
    for i, (hist, range_fit) in enumerate(zip(hists, config["fit"]["range"])):
        mean: float = hist.GetMean()
        sigma: float = hist.GetStdDev()

        func: r.TF1 = r.TF1("func", "gaus", *range_fit, 3)
        func.SetParameter(1, mean)
        func.SetParameter(2, sigma)

        fit_res = hist.Fit(func, "RS")

        hfit_res = r.TH1D(
            f"hFitRes{hist.GetName()}",
            "",
            len(labels_hfit_res),
            0,
            len(labels_hfit_res) - 1,
        )

        my_res = [
            fit_res.Status(),
            fit_res.Chi2(),
            fit_res.Ndf(),
            fit_res.Parameter(0),
            fit_res.Parameter(1),
            fit_res.Parameter(2),
            *range_fit,
        ]
        errors = [
            0,  # no error on fit status
            0,  # no error on chi2
            0,  # no error on ndf
            fit_res.ParError(0),
            fit_res.ParError(1),
            fit_res.ParError(2),
            0,  # no error on xmin
            0,  # no error on xmax
        ]
        for ilabel, (label, res, error) in enumerate(
            zip(labels_hfit_res, my_res, errors)
        ):
            ibin = ilabel + 1
            hfit_res.GetXaxis().SetBinLabel(ibin, label)
            hfit_res.SetBinContent(ibin, res)
            hfit_res.SetBinError(ibin, error)

        hfit_results.append(hfit_res)

    # save in output file
    for hist in hists:
        name = hist.GetName()
        hist.SetName(f"QA_{name}")
        hist.Write()
    for hfit_res in hfit_results:
        hfit_res.Write()
    outfile.Close()


if __name__ == "__main__":
    parser = ArgumentParser(description="Arguments")
    parser.add_argument("name_config_file", metavar="text", default="config.yaml")
    args = parser.parse_args()
    main(args.name_config_file)

File: Calibration/Fitter/src/fit.cxx
/// fit.cpp
/// author: Alexandre BIGOT, alexandre.bigot@iphc.cnrs.fr
/// brief
#include <RooAbsDataHelper.h>
#include <yaml-cpp/yaml.h>

#include <ROOT/RDataFrame.hxx>
#include <iostream>
#include <string>
#include <vector>

#include "ConfigHandler.h"
#include "RooAbsArg.h"
#include "RooAbsData.h"
#include "RooAddPdf.h"
#include "RooFFTConvPdf.h"
#include "RooFitResult.h"
#include "RooGaussian.h"
#include "RooLandau.h"
#include "RooPlot.h"
#include "RooRealVar.h"
#include "TApplication.h"
#include "TF1.h"
#include "TFile.h"
#include "TH1D.h"
#include "TString.h"
#include "TTree.h"

enum InitFitPars : uint8_t { kStart = 0, kMin, kMax, nInitFitPars };

const std::vector<TString> NAME_PARS = {"MPV_landau", "sigma_landau",
                                        "mu_gauss", "sigma_gauss", "norm"};
const std::vector<TString> LABELS_HFIT_RES = {
    "FitStatus", "Chi2OverNdf",  "Xmin",       "Xmax",
    "x_{MPV}",   "#sigma_{MPV}", "MPV_landau", "sigma_landau",
    "mu_gauss",  "sigma_gauss",  "norm"};

/// @brief Main function to perform the fit
/// @param argc
/// @param argv
int main(int argc, char** argv) {
    static constexpr bool debug = true;
    static constexpr bool batchMode = true;

    std::string nameCfgFile =
        "/Users/abigot/CLINM-AF/Calibration/ConfigFiles/FitRealData/"
        "config_fit_CNAO0425_Run2_Config1.yml";
    if (argc == 1) {
        std::cout << "Using default configuration file!\n";
    } else if (argc == 2) {
        nameCfgFile = static_cast<std::string>(argv[1]);
    } else {
        std::cout << "ERROR\n";
        return 1;
    }

    ConfigInput cfgInput = loadConfig<ConfigInput>(nameCfgFile);
    ConfigHistogram cfgHistogram = loadConfig<ConfigHistogram>(nameCfgFile);
    ConfigFit cfgFit = loadConfig<ConfigFit>(nameCfgFile);
    ConfigOutput cfgOutput = loadConfig<ConfigOutput>(nameCfgFile);

    // handle input
    const TString nameInfile = cfgInput.nameFile;
    const TString nameTree = cfgInput.nameTree;
    const auto nameBranches = cfgInput.nameBranches;
    // histogram config
    const auto nameHistos = cfgHistogram.names;
    const auto nbins = cfgHistogram.nbins;
    // fit
    const auto ranges = cfgFit.ranges;
    // fit pars
    const auto mpvsLandau = cfgFit.mpvsLandau;
    const auto sigmasLandau = cfgFit.sigmasLandau;
    const auto musGauss = cfgFit.musGauss;
    const auto sigmasGauss = cfgFit.sigmasGauss;
    const auto norms = cfgFit.norms;
    // output
    TString nameOutfile = "test_output.root";  // cfgOutput.nameFile;

    // Config for fit
    if (debug) {
        std::cout << "Ranges:\n";
        for (const auto& range : ranges) {
            std::cout << range[0] << " , " << range[1] << "\n";
        }
    }

    // for visualisation while using CMake
    std::unique_ptr<TApplication> app =
        std::make_unique<TApplication>("app", &argc, argv);

    if (debug) {
        std::cout << "nameInfile = " << nameInfile << "\n";
        std::cout << "nameTree = " << nameTree << "\n";
    }

    TFile* infile = new TFile(nameInfile.Data(), "read");

    TTree* tree = (TTree*)infile->Get<TTree>(nameTree.Data());

    std::vector<TH1D> hists;
    std::vector<TH1D> hFitResults;

    for (size_t iBranch = 0; iBranch < nameBranches.size(); ++iBranch) {
        if (iBranch > 0) {
            continue;
        }
        const auto nameBranch = nameBranches[iBranch];
        const auto nameData = nameHistos[iBranch];
        const auto nbin = nbins[iBranch];
        const auto rangeFit = ranges[iBranch];
        const auto initMpvLandau = mpvsLandau[iBranch];
        const auto initSigmaLandau = sigmasLandau[iBranch];
        const auto initMuGauss = musGauss[iBranch];
        const auto initSigmaGauss = sigmasGauss[iBranch];
        const auto initNorm = norms[iBranch];
        // const auto ymin = ymins[iBranch];
        // const auto ymax = ymaxs[iBranch];

        // Get dataframe
        ROOT::RDataFrame df(nameTree, infile, {nameBranch.Data()});

        // save histogram
        TString title("");
        title.Form(";%s; Entries;", nameData.Data());
        auto histTmp =
            df.Histo1D({nameData, title, nbin, rangeFit[0], rangeFit[1]},
                       nameBranch)
                ->Clone();
        auto hist = static_cast<TH1D*>(histTmp);
        hist->SetDirectory(0);
        hists.emplace_back(*hist);

        // start defining RooFit variables
        TString nameX("");
        nameX.Form("x%zu", iBranch);
        if (debug) std::cout << "nameX = " << nameX << "\n";
        RooRealVar x(nameBranch, nameX, rangeFit[0], rangeFit[1]);
        // Set #bins to be used for FFT sampling to 10000
        x.setBins(10000, "cache");
        x.setBins(nbin);

        auto frame = x.frame();

        TString nameDataset("");
        nameDataset.Form("dataset%zu", iBranch);
        auto dataset = df.Book<double>(
            RooDataSetHelper(nameDataset,         // Name
                             "Title of dataset",  // Title
                             RooArgSet(x)         // Variables in this dataset
                             ),
            {nameBranch.Data()}  // Column names in RDataFrame.
        );

        // construct landau
        TString nameMlVar("");
        nameMlVar.Form("MPV_landau%zu", iBranch);
        RooRealVar ml(nameMlVar, "MPV_landau", initMpvLandau[kStart],
                      initMpvLandau[kMin], initMpvLandau[kMax]);
        TString nameSlVar("");
        nameSlVar.Form("sigma_landau%zu", iBranch);
        RooRealVar sl(nameSlVar, "sigma_landau", initSigmaLandau[kStart],
                      initSigmaLandau[kMin], initSigmaLandau[kMax]);
        TString nameLandau("");
        nameLandau.Form("landau%zu", iBranch);
        RooLandau landau(nameLandau, "landau", x, ml, sl);

        // construct gauss
        TString nameMgVar("");
        nameMgVar.Form("mu_gauss%zu", iBranch);
        RooRealVar mg(nameMgVar, "mu_gauss", initMuGauss[kStart],
                      initMuGauss[kMin], initMuGauss[kMax]);
        mg.setConstant(true);  // fix mu Gauss
        TString nameSgVar("");
        nameSgVar.Form("sigma_gauss%zu", iBranch);
        RooRealVar sg(nameSgVar, "mu_gauss", initSigmaGauss[kStart],
                      initSigmaGauss[kMin], initSigmaGauss[kMax]);
        TString nameGauss("");
        nameGauss.Form("gauss%zu", iBranch);
        RooGaussian gauss(nameGauss, "gauss", x, mg, sg);

        // construct landau (x) gauss
        TString nameLxg("");
        nameLxg.Form("lxg%zu", iBranch);
        RooFFTConvPdf lxg(nameLxg, "landau (X) gauss", x, landau, gauss);
        // add norm
        TString nameNormVar("");
        nameNormVar.Form("norm%zu", iBranch);
        RooRealVar normVar(nameNormVar, "norm", initNorm[kStart],
                           initNorm[kMin], initNorm[kMax]);
        TString nameModel("");
        nameModel.Form("model%zu", iBranch);
        RooAddPdf model(nameModel, "model", {lxg}, {normVar});

        // fit
        auto fitres = model.fitTo(*dataset, RooFit::NumCPU(3), RooFit::Save());

        // plot
        dataset->plotOn(frame, RooFit::MarkerStyle(kFullCircle));
        model.plotOn(frame, RooFit::LineColor(kAzure + 4), RooFit::LineWidth(2),
                     RooFit::MoveToBack());
        frame->Draw();

        // ...

        auto funcModel = model.asTF(x);
        double xMpv = funcModel->GetMaximumX();
        if (debug) {
            std::cout << "\n\n\n\n\n\n";
            std::cout << "x MPV = " << xMpv << "\n";
        }
        auto mlVarFitted =
            (RooRealVar*)fitres->floatParsFinal().find(nameMlVar.Data());
        double sMpv = mlVarFitted->getAsymErrorHi();

        // histogram with fit results
        TString nameHFitRes("");
        nameHFitRes.Form("hFitRes%s", nameData.Data());
        auto hFitResult = TH1D(nameHFitRes, "", LABELS_HFIT_RES.size(), 0,
                               LABELS_HFIT_RES.size() - 1);
        std::vector<double> myRes = {static_cast<double>(fitres->status()),
                                     frame->chiSquare(),
                                     rangeFit[0],
                                     rangeFit[1],
                                     xMpv,
                                     sMpv};

        std::vector<double> errors = {0, 0, 0, 0, 0, 0};

        for (const auto& nameParOriginal : NAME_PARS) {
            if (nameParOriginal == "mu_gauss") {  // mu Gauss is fixed
                myRes.emplace_back(0);
                continue;
            }
            TString namePar("");
            namePar.Form("%s%zu", nameParOriginal.Data(), iBranch);
            auto parFitted =
                (RooRealVar*)fitres->floatParsFinal().find(namePar.Data());
            myRes.emplace_back(parFitted->getValV());
            errors.emplace_back(parFitted->getAsymErrorHi());
        }

        for (size_t iLabel = 0; iLabel < myRes.size(); iLabel++) {
            int ibin = iLabel + 1;
            hFitResult.GetXaxis()->SetBinLabel(ibin, LABELS_HFIT_RES[iLabel]);
            hFitResult.SetBinContent(ibin, myRes[iLabel]);
            hFitResult.SetBinError(ibin, errors[iLabel]);
        }

        if (debug) {
            std::cout << "\n\nFit result: " << hFitResult.GetBinContent(1)
                      << "\n\n";
        }

        hFitResults.emplace_back(hFitResult);
    }

    infile->Close();

    TFile* outfile = new TFile(nameOutfile.Data(), "recreate");
    for (auto& hist : hists) {
        hist.Write();
    }
    for (auto& hFitResult : hFitResults) {
        hFitResult.Write();
    }

    outfile->Close();

    // std::string nameBranch = "pl0ntu.fPlAmplitude";
    // // ROOT::RDataFrame dfInitial(*tree);
    // ROOT::RDataFrame df(nameTree.Data(), infile, {nameBranch.data()});

    // const bool useRooFit = false;

    // const int nbins = 100;

    // // using Root(!)Fit
    // if (!useRooFit) {
    //     auto histTmp =
    //         df.Histo1D({"fDeltaEPl1", "", nbins, 40., 120.},
    //         nameBranch.data())
    //             ->Clone();
    //     auto hist = static_cast<TH1D*>(histTmp);
    //     hist->Draw();
    // }

    // // using RooFit
    // if (useRooFit) {
    //     RooRealVar x("x", "x", 40., 120.);
    //     x.setBins(nbins);

    //     auto dataset = df.Book<double>(
    //         RooDataSetHelper("dataset",           // Name
    //                          "Title of dataset",  // Title
    //                          RooArgSet(x)         // Variables in this
    //                          dataset
    //                          ),
    //         {nameBranch.data()}  // Column names in RDataFrame.
    //     );
    // }

    if (!batchMode) {
        app->Run();
    }

    return 0;
}